src/vset.h

```diff
@@ -0,0 +1,317 @@
+ *
+ *-----------------------------------------------------------------------------
+ * Entry Type Contract
+ *-----------------------------------------------------------------------------
+ *
+ * Users must supply a `volatileEntryType` implementation:
+ *
+ *     typedef struct {
+ *         sds (*entryGetKey)(const void *entry);      // get key
+ *         long long (*getExpiry)(const void *entry);  // get expiry
+ *         int (*expire)(void *db, void *o, void *entry); // trigger expiry
+ *     } volatileEntryType;
+ *
+ *-----------------------------------------------------------------------------
+ * Public API
+ *-----------------------------------------------------------------------------
+ *
+ * Create/Free:
+ *     void vsetInit(vset *set);
+ *     void vsetClear(vset *set);
+ *
+ * Mutation:
+ *     bool vsetAddEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+ *     bool vsetRemoveEntry(vset *set, vsetGetExpiryFunc getExpiry, void *entry);
+ *     bool vsetUpdateEntry(vset *set, vsetGetExpiryFunc getExpiry, void *old_entry,
+ *                                void *new_entry, long long old_expiry,
+ *                                long long new_expiry);
+ *
+ * Expiry Retrieval/Removal:
+ *     long long vsetEstimatedEarliestExpiry(vset *set, vsetGetExpiryFunc getExpiry);
+ *     size_t vsetPopExpired(vset *set, vsetGetExpiryFunc getExpiry, vsetExpiryFunc expiryFunc, mstime_t now, size_t max_count, void *ctx);
+ *
+ * Utilities:
+ *     bool vsetIsEmpty(vset *set);
+ *
+ * Iteration:
+ *     void vsetStart(vset *set, vsetIterator *it);
+ *     bool vsetNext(vsetIterator *it, void **entryptr);
+ *     void vsetStop(vsetIterator *it);
+ *
+ *-----------------------------------------------------------------------------
+ * Iteration Support
+ *-----------------------------------------------------------------------------
+ *
+ * Iterator structure maintains context across all bucket types:
+ *
+ *     typedef struct vsetIterator {
+ *         raxIterator riter;           // for RAX
+ *         hashtableIterator hiter;     // for HT
+ *         uint32_t viter;              // for VECTOR
+ *         void *vsingle;               // for SINGLE
+ *         vsetBucket *parent_bucket;   // owning bucket
+ *         vsetBucket *bucket;          // active bucket
+ *         void *entry;                 // current entry
+ *         long long bucket_ts;         // for RAX
+ *         int iteration_state;         // internal FSM
+ *     } vsetIterator;
+ * */
+
+#define VOLATILESET_BUCKET_INTERVAL_MAX (1LL << 13LL) // 2^13 = 8192 milliseconds
+#define VOLATILESET_BUCKET_INTERVAL_MIN (1LL << 4LL)  // 2^4 = 16 milliseconds
+
+#define VOLATILESET_VECTOR_BUCKET_MAX_SIZE 127
+
+typedef long long (*vsetGetExpiryFunc)(const void *entry);
+typedef int (*vsetExpiryFunc)(void *entry, void *ctx);
+
+// Generic bucket type
+typedef void vsetBucket;
+
+// vset is just a pointer to a bucket
+typedef vsetBucket *vset;
```

Comment:
I don't think you want the public `vset` to be defined in terms of a private `vsetBucket`.  I'd recommend moving `vsetBucket` to the .c file, and define `vset` opaquely like `typedef struct vset vset;`.

Note:
By forward declaring `typedef struct vset vset;`, the compiler knows that vset is a struct, but doesn't yet know how big it is.  It's then legal to create pointer variables like `vset *myVset;` or use it as a function parameter.

Then later, in the .c file, you can do `struct vset { ... };`